File: l2-loginserver/src/main/java/com/vvygulyarniy/l2/loginserver/netty/NettyLoginServerHandler.java
package com.vvygulyarniy.l2.loginserver.netty;

import com.l2server.network.clientpackets.login.L2LoginClientPacket;
import com.l2server.network.login.L2LoginClient;
import com.l2server.network.serverpackets.login.Init;
import com.vvygulyarniy.l2.loginserver.LoginController;
import com.vvygulyarniy.l2.loginserver.logic.LoginPacketsProcessor;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.util.AttributeKey;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class NettyLoginServerHandler extends ChannelInboundHandlerAdapter {
    private final AttributeKey<L2LoginClient> clientKey = AttributeKey.valueOf("l2LoginClient");
    private final LoginController loginController;
    private final LoginPacketsProcessor packetProcessor;

    public NettyLoginServerHandler(LoginController loginController, LoginPacketsProcessor packetProcessor) {
        this.loginController = loginController;
        this.packetProcessor = packetProcessor;
    }

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        L2LoginClient client = new L2LoginClient(new NettyLoginClientConnection(ctx),
                                                 loginController.getScrambledRSAKeyPair(),
                                                 loginController.getBlowfishKey());
        ctx.channel().attr(clientKey).set(client);
        log.info("Channel registered: {}", ctx);
    }

    @Override
    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
        ctx.channel().attr(clientKey).set(null);
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        log.info("Got message: {}", msg);
        L2LoginClient client = ctx.channel().attr(clientKey).get();
        L2LoginClientPacket packet = (L2LoginClientPacket) msg;
        packet.process(packetProcessor, client);
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        log.info("Channel active: {}", ctx);
        L2LoginClient client = ctx.channel().attr(clientKey).get();
        Init initPacket = new Init(client);


        ctx.channel().writeAndFlush(initPacket);
    }


}

File: l2-loginserver/src/main/java/com/vvygulyarniy/l2/loginserver/netty/NettyLoginServerHandler.kt
package com.vvygulyarniy.l2.loginserver.netty

import com.l2server.crypt.LoginCrypt
import com.l2server.network.ClientPacket
import com.l2server.network.communication.NettyClientCommunicator
import com.vvygulyarniy.l2.loginserver.communication.CommunicationManager
import com.vvygulyarniy.l2.loginserver.communication.packet.server.Init
import com.vvygulyarniy.l2.loginserver.events.router.ClientPacketsRouter
import com.vvygulyarniy.l2.loginserver.session.SessionIdFactory
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class NettyLoginServerHandler(private val communicationManager: CommunicationManager,
                              private val sessionFactory: SessionIdFactory,
                              private val packetProcessor: ClientPacketsRouter) : ChannelInboundHandlerAdapter() {

    override fun channelRegistered(ctx: ChannelHandlerContext) {
        val crypt = LoginCrypt()
        val sessionId = sessionFactory.generate()
        ctx.setSessionId(sessionId)
        ctx.setCrypt(crypt)

        communicationManager.newChannel(sessionId, NettyClientCommunicator(ctx))
        log.info("Channel registered: {}", ctx)
    }

    @Throws(Exception::class)
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        log.info("Got message: {}", msg)
        val packet = msg as ClientPacket
        packetProcessor.handlePacket(ctx.getSessionId(), packet)
    }

    @Throws(Exception::class)
    override fun channelActive(ctx: ChannelHandlerContext) {
        log.info("Channel active: {}", ctx)
        val crypt = ctx.getCrypt()

        val initPacket = Init(crypt.scrambledKeyPair.modulus, crypt.blowfishKey, ctx.getSessionId().toInt())

        communicationManager.sendPacket(ctx.getSessionId(), initPacket)
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext?, cause: Throwable?) {
        log.error("Exception got while processing user request", cause)
        ctx?.let { communicationManager.closeChannel(it.getSessionId()) }
    }

    companion object {
        val log: Logger = LoggerFactory.getLogger(NettyLoginServerHandler::class.java)
    }

}
